pages/posts/supporting-polymorphism.md

```diff
@@ -0,0 +1,431 @@
+    "name": { "type": "string" }
+  },
+  "required": [ "motorCount", "name" ]
+}
+
+{
+  "id": "car",
+  "allOf": [
+    { "$ref": "vehicle" }
+  ],
+  "properties": {
+    "doorCount": { "type": "integer" },
+    "rangeInMiles": { "type": "number" },
+  },
+  "required": [ "doorCount", "rangeInMiles" ]
+}
+```
+
+It doesn't seem like the type system is impacted much, if at all.  If I were to get JSON data that met the requirements of `boat`, it would also meet the requirements of `vehicle`.  The "is-a" relationship still works fine.
+
+However, I think the reduced usability of not being able to omit properties hinders this approach enough that having `required` isn't worth it.
+
+#### Disallowing Undeclared Properties
+
+Disallowing undeclared properties serves two purposes.
+
+First, we tend not to want junk in our data.  This also attempts to align the subtractive constraints model available in JSON Schema to the additive type definition models that are present in programming languages.  JSON Schema starts with "everything is valid" and narrows the valid space by adding constraints.  By disallowing _all_ undeclared properties, we, in effect, move (somewhat) to an additive model where declaring a property expands the valid space.
+
+Second, JSON data (and YAML, if that's your thing) is often written by humans.  Humans make mistakes.  By disallowing undeclared properties, we protect against mistakes like mistyping property names.  (This kind of protection is a surprisingly common request.)
+
+There are two keywords that are typically cited to help, and they work subtly differently: `additionalProperties` and `unevaluatedProperties`.
+
+`additionalProperties` works by validating any properties left behind by `properties` and `patternProperties`.   The limitation is that it can't "look inside" of subschemas defined in keywords like `allOf`, `anyOf`, or `$ref`.  That means that this schema
+
+```json
+{
+  "allOf": [
+    {
+      "properties": {
+        "foo": true
+      }
+    }
+  ],
+  "additionalProperties": false
+}
+```
+
+will fail any object with a property, even if that property is `foo`.  This presents a problem for our derived models because some of the properties we want to allow are defined in another schema (i.e. `vehicle`).
+
+To support this "looking inside" other keywords behavior, we introduced `unevaluatedProperties` with draft 2019-09.  `unevaluatedProperties` does exactly the same thing as `additionalProperties` except that it can "look inside" subschemas of the schema object to which it belongs.  So this schema
+
+```json
+{
+  "allOf": [
+    {
+      "properties": {
+        "foo": true
+      }
+    }
+  ],
+  "unevaluatedProperties": false
+}
+```
+
+will properly allow objects with a `foo` property because it can "see" the `foo` property declaration inside the subschema in `allOf`.
+
+So let's add `"unevaluatedProperties": false` to our schemas and see what happens.
+
+```json
+{
+  "id": "vehicle",
```

Comment:
Did you mean to write this with draft-04 `id`?

Reply:
haha, no.